File: backend/csvimporter/src/main/java/com/simra/app/csvimporter/mapmatching/MapMatchingService.java
package com.simra.app.csvimporter.mapmatching;

import com.graphhopper.GraphHopper;
import com.graphhopper.PathWrapper;
import com.graphhopper.matching.MatchResult;
import com.graphhopper.matching.Observation;
import com.graphhopper.reader.osm.GraphHopperOSM;
import com.graphhopper.routing.AlgorithmOptions;
import com.graphhopper.routing.util.FlagEncoder;
import com.graphhopper.routing.util.HintsMap;
import com.graphhopper.routing.util.TraversalMode;
import com.graphhopper.routing.weighting.FastestWeighting;
import com.graphhopper.util.*;
import com.graphhopper.util.shapes.GHPoint;
import com.graphhopper.util.shapes.GHPoint3D;
import com.simra.app.csvimporter.model.RideCSV;
import com.simra.app.csvimporter.service.Utils;
import io.jenetics.jpx.GPX;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Service
public class MapMatchingService {

    private GraphHopper graphHopper;

    public MapMatchingService() {

        CmdArgs graphHopperConfiguration = new CmdArgs();
        graphHopperConfiguration.put("graph.flag_encoders", "bike");
        graphHopperConfiguration.put("datareader.file", "backend/csvimporter/map-data/Brandenburg_and_Berlin.osm.pbf");

        this.graphHopper = new GraphHopperOSM().init(graphHopperConfiguration);
        this.graphHopper.importOrLoad();
    }

    private Float currentRouteDistance = 0F;

    private Long currentRouteDuration = 0L;


    public Float getCurrentRouteDistance() {
        return currentRouteDistance;
    }

    public Long getCurrentRouteDuration() {
        return currentRouteDuration;
    }

    public void setCurrentRouteDistance(Float currentRouteDistance) {
        this.currentRouteDistance = currentRouteDistance;
    }

    /**
     * Snaps the GPS Coordinates onto OSM-Streets.
     * TODO Add possibility to snap Routes in other cities than Berlin
     */
    public List<RideCSV> matchToMap(List<RideCSV> rideBeans) {

        GPX gpx = GPX.builder()
                .addTrack(track -> track
                        .addSegment(segment ->
                                rideBeans.forEach(ride ->
                                        segment.addPoint(p -> p.lat(Double.parseDouble(ride.getLat())).lon(Double.parseDouble(ride.getLon()))))))
                .build();

        StopWatch matchSW = new StopWatch();

        Translation tr = new TranslationMap().doImport().getWithFallBack(Locale.GERMANY);


        List<Observation> measurements = gpx.getTracks().get(0).segments().flatMap(segment ->
                segment.points().map(point ->
                        new Observation(new GHPoint(point.getLatitude().doubleValue(), point.getLongitude().doubleValue())))).collect(Collectors.toList());

        FlagEncoder firstEncoder = graphHopper.getEncodingManager().fetchEdgeEncoders().get(0);
        AlgorithmOptions opts = AlgorithmOptions.start()
                .algorithm(Parameters.Algorithms.DIJKSTRA_BI)
                .traversalMode(TraversalMode.EDGE_BASED)
                .weighting(new FastestWeighting(firstEncoder))
                .maxVisitedNodes(2000)
                .hints(new HintsMap().put("weighting", "fastest")
                        .put("vehicle", "bike"))
                // Penalizing inner-link U-turns only works with fastest weighting, since
                // shortest weighting does not apply penalties to unfavored virtual edges.
                .build();

        CustomMapMatching mapMatching = new CustomMapMatching(graphHopper, new HintsMap(opts.getHints())) {

        };
        mapMatching.setTransitionProbabilityBeta(0.5);
        mapMatching.setMeasurementErrorSigma(50.0);


        PathMerger pathMerger = new PathMerger(graphHopper.getGraphHopperStorage(), opts.getWeighting());

        matchSW.start();
        MatchResult mr = mapMatching.doWork(measurements);
        matchSW.stop();

        PathWrapper pathWrapper = new PathWrapper();
        pathMerger.doWork(pathWrapper, Arrays.asList(mr.getMergedPath()), graphHopper.getEncodingManager(), tr);

        // Copy metainfo of closest (Raw) Point to Snapped Point

        List<RideCSV> result = StreamSupport.stream(pathWrapper.getPoints().spliterator(), false)
                .map(point -> {
                            RideCSV rideCSV = new RideCSV();
                            rideCSV.setLat(String.valueOf(point.lat));
                            rideCSV.setLon(String.valueOf(point.lon));
                            RideCSV nextPoint = findNearestPoint(rideBeans, point);
                            rideCSV.setA(nextPoint.getA());
                            rideCSV.setB(nextPoint.getB());
                            rideCSV.setC(nextPoint.getC());
                            rideCSV.setAcc(nextPoint.getAcc());
                            rideCSV.setTimeStamp(nextPoint.getTimeStamp());
                            rideCSV.setX(nextPoint.getX());
                            rideCSV.setY(nextPoint.getY());
                            rideCSV.setZ(nextPoint.getZ());
                            return rideCSV;
                        }
                ).collect(Collectors.toList());

        currentRouteDistance = (float) pathWrapper.getDistance();
        currentRouteDuration = rideBeans.get(rideBeans.size() - 1).getTimeStamp() - rideBeans.get(0).getTimeStamp();

        return result;
    }

    private RideCSV findNearestPoint(List<RideCSV> rawRideCSVList, GHPoint3D matchedPoint) {
        AtomicReference<RideCSV> nextPoint = new AtomicReference<>();
        AtomicReference<Double> minDist = new AtomicReference<>(Double.MAX_VALUE);
        rawRideCSVList.forEach(it -> {
            double dist = Utils.calcDistance(Double.parseDouble(it.getLat()), Double.parseDouble(it.getLon()), matchedPoint.lat, matchedPoint.lon);
            if (dist < minDist.get()) {
                minDist.set(dist);
                nextPoint.set(it);
            }
        });
        return nextPoint.get();
    }
}

File: backend/csvimporter/src/main/java/com/simra/app/csvimporter/mapmatching/MapMatchingService.kt
package main.java.com.simra.app.csvimporter.mapmatching

import com.graphhopper.PathWrapper
import com.graphhopper.matching.MapMatching
import com.graphhopper.matching.Observation
import com.graphhopper.reader.osm.GraphHopperOSM
import com.graphhopper.routing.AlgorithmOptions
import com.graphhopper.routing.util.HintsMap
import com.graphhopper.routing.util.TraversalMode
import com.graphhopper.routing.weighting.FastestWeighting
import com.graphhopper.util.*
import com.graphhopper.util.shapes.GHPoint
import com.graphhopper.util.shapes.GHPoint3D
import io.jenetics.jpx.GPX
import main.java.com.simra.app.csvimporter.model.RideCSV
import java.util.*
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

class MapMatchingService {

    var currentRouteDistance: Float = 0F;

    /**
     * Snaps the GPS Coordinates onto OSM-Streets.
     * TODO Add possibility to snap Routes in other cities than Berlin
     */
    fun matchToMap(rideBeans: List<RideCSV>): List<RideCSV> {

        val gpx = GPX.builder()
                .addTrack { track ->
                    track.addSegment { segment ->
                        rideBeans.forEach {
                            segment.addPoint { p -> p.lat(it.lat.toDouble()).lon(it.lon.toDouble()) }
                        }
                    }
                }.build()


        val graphHopperConfiguration = CmdArgs()
        graphHopperConfiguration.put("graph.flag_encoders", "bike")
        graphHopperConfiguration.put("datareader.file", "backend/csvimporter/map-data/Brandenburg_and_Berlin.osm.pbf")

        val graphHopper = GraphHopperOSM().init(graphHopperConfiguration)
        graphHopper.importOrLoad()

        val firstEncoder = graphHopper.encodingManager.fetchEdgeEncoders()[0]
        val opts = AlgorithmOptions.start()
                .algorithm(Parameters.Algorithms.DIJKSTRA_BI)
                .traversalMode(TraversalMode.EDGE_BASED)
                .weighting(FastestWeighting(firstEncoder))
                .maxVisitedNodes(2000)
                .hints(HintsMap().put("weighting", "fastest")
                        .put("vehicle", "bike"))
                // Penalizing inner-link U-turns only works with fastest weighting, since
                // shortest weighting does not apply penalties to unfavored virtual edges.
                .build()

        val mapMatching = MapMatching(graphHopper, HintsMap(opts.hints))
        mapMatching.setTransitionProbabilityBeta(2.0)
        mapMatching.setMeasurementErrorSigma(50.0)

        val matchSW = StopWatch()

        val tr = TranslationMap().doImport().getWithFallBack(Locale.GERMANY)

        val measurements = gpx.tracks[0].segments.flatMap { segment -> segment.map { Observation(GHPoint(it.latitude.toDouble(), it.longitude.toDouble())) } }
        matchSW.start()
        val mr = mapMatching.doWork(measurements)
        matchSW.stop()
        println("\tmatches:\t" + mr.edgeMatches.size + ", gps entries:" + measurements.size)


        val pathWrapper = PathWrapper()
        PathMerger(graphHopper.graphHopperStorage, opts.weighting).doWork(pathWrapper, listOf(mr.mergedPath), graphHopper.encodingManager, tr)

        // Copy metainfo of closest (Raw) Point to Snapped Point
        val result = pathWrapper.points.map {
            val rideCSV = RideCSV()

            rideCSV.lat = it.lat.toString()
            rideCSV.lon = it.lon.toString()

            val nextPoint = findNearestPoint(rideBeans, it)
            rideCSV.a = nextPoint.a
            rideCSV.b = nextPoint.b
            rideCSV.c = nextPoint.c
            rideCSV.acc = nextPoint.acc
            rideCSV.timeStamp = nextPoint.timeStamp
            rideCSV.x = nextPoint.x
            rideCSV.y = nextPoint.y
            rideCSV.z = nextPoint.z

            rideCSV
        }

        currentRouteDistance = pathWrapper.distance.toFloat()

        println("matching took: " + matchSW.seconds)

        return result
    }

    private fun findNearestPoint(rawRideCSVList: List<RideCSV>, matchedPoint: GHPoint3D): RideCSV {
        var nextPoint: RideCSV? = null
        var minDist: Double? = null
        rawRideCSVList.forEach {
            val dist = distance(it.lat.toDouble(), it.lon.toDouble(), matchedPoint.lat, matchedPoint.lon)
            if (minDist == null || dist < minDist!!) {
                minDist = dist
                nextPoint = it
            }
        }
        return nextPoint!!
    }

    private fun distance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val R = 6371 // Radius of the earth

        val latDistance = Math.toRadians(lat2 - lat1)
        val lonDistance = Math.toRadians(lon2 - lon1)
        val a = sin(latDistance / 2) * sin(latDistance / 2) + (cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2))
                * sin(lonDistance / 2) * sin(lonDistance / 2))
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))

        return R.toDouble() * c * 1000.0
    }
}